compiler/decls.go

```diff
@@ -0,0 +1,541 @@
+	// Go compiler/linker toolchain. Used by GopherJS to support go:linkname
+	// directives. Must be set for decls that are supported by go:linkname
+	// implementation.
+	LinkingName SymName
+	// A list of package-level JavaScript variable names this symbol needs to declare.
+	Vars []string
+	// NamedRecvType is method named recv declare.
+	NamedRecvType string
+	// JavaScript code that declares basic information about a symbol. For a type
+	// it configures basic information about the type and its identity. For a function
+	// or method it contains its compiled body.
+	DeclCode []byte
+	// JavaScript code that initializes reflection metadata about type's method list.
+	MethodListCode []byte
+	// JavaScript code that initializes the rest of reflection metadata about a type
+	// (e.g. struct fields, array type sizes, element types, etc.).
+	TypeInitCode []byte
+	// JavaScript code that needs to be executed during the package init phase to
+	// set the symbol up (e.g. initialize package-level variable value).
+	InitCode []byte
+	// Symbol's identifier used by the dead-code elimination logic, not including
+	// package path. If empty, the symbol is assumed to be alive and will not be
+	// eliminated. For methods it is the same as its receiver type identifier.
+	DceObjectFilter string
+	// The second part of the identifier used by dead-code elimination for methods.
+	// Empty for other types of symbols.
+	DceMethodFilter string
+	// List of fully qualified (including package path) DCE symbol identifiers the
+	// symbol depends on for dead code elimination purposes.
+	DceDeps []string
+	// Set to true if a function performs a blocking operation (I/O or
+	// synchronization). The compiler will have to generate function code such
+	// that it can be resumed after a blocking operation completes without
+	// blocking the main thread in the meantime.
+	Blocking bool
+}
+
+// minify returns a copy of Decl with unnecessary whites paces removed from the
+// JS code.
+func (d Decl) minify() Decl {
+	d.DeclCode = removeWhitespace(d.DeclCode, true)
+	d.MethodListCode = removeWhitespace(d.MethodListCode, true)
+	d.TypeInitCode = removeWhitespace(d.TypeInitCode, true)
+	d.InitCode = removeWhitespace(d.InitCode, true)
+	return d
+}
+
+// topLevelObjects extracts package-level variables, functions and named types
+// from the package AST.
+func (fc *funcContext) topLevelObjects(srcs sources) (vars []*types.Var, functions []*ast.FuncDecl, typeNames []*types.TypeName) {
+	if !fc.IsRoot() {
+		panic(bailout(fmt.Errorf("functionContext.discoverObjects() must be only called on the package-level context")))
+	}
+
+	for _, file := range srcs.Files {
+		for _, decl := range file.Decls {
+			switch d := decl.(type) {
+			case *ast.FuncDecl:
+				sig := fc.pkgCtx.Defs[d.Name].(*types.Func).Type().(*types.Signature)
+				if sig.Recv() == nil {
+					fc.objectName(fc.pkgCtx.Defs[d.Name]) // register toplevel name
```

Comment:
I see here (and elsewhere) that this method modifies `fc`. Is it wise to describe this in the godoc comment as well?  As currently documented, this looks like a pure function.

Edit: I see the same happening in a number of other methods, too.  It may be unnecessarily tedious to document every instance of a method that modifies the receiver here.

Reply:
Yeah, pretty much every `functionContext` method may mutate the receiver. I tired to keep functions that can be pure separate.

I'm not too happy with the state of `functionContext` type in general. Side-effect mutations are one thing, but it's also a mostly untestable god object, it directly mutates fields of a bunch of data structures instead of using a semantic API and its name is misleading — it's more like a "JS scope context", rather than "a particular function context". But so far I've got no idea how to fix all of that, so we are going to have to live with it until someone has a 💡 moment.